Hoàn thành backend

phamdinhlam105 opened the pull request.

phamdinhlam105 committed:
tạo project backend

database 3 bảng : article, category, image

phamdinhlam105 committed:
cài validator + typeorm

phamdinhlam105 committed:
fix lỗi service và đổi vị trí package

commit trước cài package nhầm folder

phamdinhlam105 committed:
fix lỗi sửa tên trong các service

phamdinhlam105 committed:
fix các lỗi logic trong service

phamdinhlam105 committed:
Merge pull request #3 from phamdinhlam105/main

merge main to back-end

phamdinhlam105 committed:
fix lỗi các service

phamdinhlam105 committed:
cài đặt ValidationPipe

NguyenNgocHongHan commented on back-end/src/app.module.ts:
```diff
@@ -0,0 +1,25 @@
+      database:'abc_intern',
+      entities: [ImageEntity,ArticleEntity,CategoryEntity],
+      synchronize:true
+    })],
```
mấy thông tin trong TypeOrmModule.forRoot({
      type: 'postgres',
      host:'localhost',
      port:5432,
      username: 'postgres',
      password:'123456',
      database:'abc_intern',
      entities: [ImageEntity,ArticleEntity,CategoryEntity],
      synchronize:true
    })], sau này mình làm trong mấy dự án nó là thông tin private nên anh thêm nó vô file .env nha, trong file back-end/.gitignore and đã thêm .env ở trỏng thì nó sẽ không đưa lên git

NguyenNgocHongHan commented on back-end/src/article-module/article.entity.ts:
```diff
@@ -0,0 +1,31 @@
+    id: UUID;
+
+    @Column({ type: 'text', nullable: false })
+    title: string;
```
mấy cột mà nội dung không dài quá ví dụ như title, thumbnail,... anh để dạng varchar 225 để giới hạn ký tự nha, ví dụ @Column({type: 'varchar', length: 225})

NguyenNgocHongHan commented on back-end/src/article-module/article.dto.ts:
```diff
@@ -0,0 +1,31 @@
+
+    @IsOptional()
+    @IsString()
+    createDate: string;
```
createDate anh không cần để trong dto nha, để nó tự sinh ra mỗi khi tạo một record mới luôn

NguyenNgocHongHan commented on back-end/src/article-module/article.entity.ts:
```diff
@@ -0,0 +1,31 @@
+    describe: string;
+
+    @Column({ type: 'timestamp', nullable: false })
+    createDate: Date;
```
anh update decorator column thành @CreateDateColumn({ type: 'timestamp without time zone', default: () => 'CURRENT_TIMESTAMP' }) để nó auto sinh ra mỗi khi có record mới, không cần truyền createDate ở dto

NguyenNgocHongHan commented on back-end/src/article-module/article.entity.ts:
```diff
@@ -0,0 +1,31 @@
+    @ManyToOne(() => CategoryEntity, (category) => category.posts)
+    category: CategoryEntity;
+
+    @OneToMany(() => ImageEntity, (image) => image.article, { onDelete: 'RESTRICT' })
```
nếu anh muốn xoá được row trong bảng con khi có khoá ngoại thì thay RESTRICT thành CASCADE nha

NguyenNgocHongHan commented on back-end/src/article-module/article.interface.ts:
```diff
@@ -0,0 +1,14 @@
+
+
+export default interface Article {
+    id: UUID;
```
**id: UUID;** ==> **id:string;**

NguyenNgocHongHan commented on back-end/src/image-module/image.ulti.ts:
```diff
@@ -0,0 +1,13 @@
+import entityToArticle from "src/article-module/article.ulti";
```
**ulti** đổi thành **util** nha anh, tương tự anh đổi tên mấy file util khác luôn nha

NguyenNgocHongHan commented on back-end/src/article-module/article.service.ts:
```diff
@@ -0,0 +1,102 @@
+        private categoryRepository: Repository<CategoryEntity>
+    ) { }
+
+    private idValid(id: string) {
```
thường tên function sẽ là 1 action thay vì danh từ, em recomment nên đổi tên thành checkIdIsValid hoặc tên khác tương tự nha

NguyenNgocHongHan commented on back-end/src/article-module/article.service.ts:
```diff
@@ -0,0 +1,102 @@
+            throw new HttpException('invalid id', HttpStatus.BAD_REQUEST);
+    }
+
+    private dateFormatValid(date: string) {
```
tương tự hàm này cũng nên đổi tên nha anh

NguyenNgocHongHan commented on back-end/src/article-module/article.service.ts:
```diff
@@ -0,0 +1,102 @@
+            throw new HttpException('invalid date format', HttpStatus.BAD_REQUEST);
+    }
+
+    async createArticle(article: CreateArticleDto): Promise<Article> {
```
để tên article cho dto dễ gây nhầm lẫn, em suggest đổi lại thành createArticleDto: CreateArticleDto

NguyenNgocHongHan commented on back-end/src/article-module/article.service.ts:
```diff
@@ -0,0 +1,102 @@
+
+    async createArticle(article: CreateArticleDto): Promise<Article> {
+        const newArticle = new ArticleEntity();
+        const { createDate, idCategory, ...rest } = article;
```
tương tự vì để article nên khi nhìn vào sẽ nghĩ là bài viết đã có trong DB, không phải là param truyền vào

NguyenNgocHongHan commented on back-end/src/article-module/article.service.ts:
```diff
@@ -0,0 +1,102 @@
+        newArticle.category = category;
+        if (createDate)
+            this.dateFormatValid(createDate);
+        newArticle.createDate = createDate ? new Date(createDate) : new Date();
```
bỏ createDate nha anh

phamdinhlam105 committed:
resolve các comment thêm soft delete

phamdinhlam105 committed:
sửa lại return của delete

NguyenNgocHongHan commented on back-end/src/image-module/image.util.ts:
```diff
@@ -0,0 +1,15 @@
+import { ImageEntity } from "./image.entity";
+import Image from './image.interface';
+
+export default function entityToImage(imageEntity: ImageEntity): Image {
```
file util thường chứa các function để thực hiện hành vi gì đó, ví dụ tiêu biểu là convert hoặc format,... nên tên function anh đổi tên lại dạng động từ nha (vd: convertImageEntity,...)

NguyenNgocHongHan commented on back-end/src/image-module/image.interface.ts:
```diff
@@ -0,0 +1,9 @@
+import Article from "src/article-module/article.interface";
+
+export default interface Image {
```
tên của interface nên bắt đầu là I[...], ví dụ: IImage, còn nếu anh export dang type thì sẽ thành: export type TImage = {...}

phamdinhlam105 committed:
resolve commit, bỏ expose trong dto

thay expose + plainToClass bằng whitelist và UpdateDto. whitelist trong controller dùng để filter các properties
